README.md
usb-creator-patches
===================

Why I do it? Because my eyes can't see this poverty...


debian/changelog
usb-creator (0.2.50) saucy; urgency=low

  * Remove untranslated and unnecessary string

 -- Jonathan Riddell <jriddell@ubuntu.com>  Fri, 27 Sep 2013 15:48:29 +0100

usb-creator (0.2.49) saucy; urgency=low

  * SECURITY UPDATE: possible privilege escalation via policykit UID lookup
    race.
    - bin/usb-creator-helper: pass system-bus-name as a subject instead of
      pid so policykit can get the information from the system bus.
    - CVE-2013-1063

 -- Marc Deslauriers <marc.deslauriers@ubuntu.com>  Wed, 18 Sep 2013 13:23:48 -0400

usb-creator (0.2.48) saucy; urgency=low
 
  * Depend on gir1.2-gudev-1.0 (LP: #1178057)

 -- Jeremy Bicha <jbicha@ubuntu.com>  Fri, 14 Jun 2013 13:03:21 -0400

usb-creator (0.2.47) raring; urgency=low

  * Purge the upstart job.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Wed, 20 Mar 2013 09:01:26 +0000

usb-creator (0.2.46) raring; urgency=low

  * Simplify upstart job, do checks in start on conditions and not in
    pre-start script.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Wed, 06 Feb 2013 19:47:52 +0000

usb-creator (0.2.45.1) raring; urgency=low

  * KDE frontend python3 fixes from Riddell.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Mon, 28 Jan 2013 13:35:56 +0000

usb-creator (0.2.45) raring; urgency=low

  * Support flashing devices using fastboot (e.g. Nexus7).
  * Update translations.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Mon, 28 Jan 2013 12:51:45 +0000

usb-creator (0.2.44) raring; urgency=low

  * Revert sync remount flag. It's too slow.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Thu, 17 Jan 2013 18:01:22 +0000

usb-creator (0.2.43) raring; urgency=low

  * Port helper script to gobject-interspection. (LP: #1078810)
  * Fix pygi depreciation warnings.
  * Fix passing one to many arguments to one Gtk.MessageDialog.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Thu, 10 Jan 2013 14:58:48 +0000

usb-creator (0.2.42) raring; urgency=low

  * Restrict architectures to i386 & amd64. Although it's all pure python,
    it depends on executing syslinux which is only available on i386 &
    amd64.

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Mon, 26 Nov 2012 11:02:30 +0000

usb-creator (0.2.41) raring; urgency=low

  * Switch packaging to use python3.
  * Drop utf-8/gnome workaround, now fixed in gnome.
  * Do not use INT_MAX (LP: #1076305).

 -- Dmitrijs Ledkovs <dmitrij.ledkov@ubuntu.com>  Fri, 31 Aug 2012 16:52:34 +0100

usb-creator (0.2.40ubuntu1) quantal; urgency=low

  * debian/control: remove Build-Depends on kdesdk-scripts so kdesdk
    can live in universe

 -- Jamie Strandboge <jamie@ubuntu.com>  Tue, 31 Jul 2012 11:02:49 -0500

usb-creator (0.2.40) quantal; urgency=low

  [ Dmitrijs Ledkovs ]
  * Increase the height of both source and target treeviews in the GTK+
    frontend. Thanks to Ross Lagerwall.
  * Search Desktop in addition to Downloads directory for ISOs. Thanks to
    Lucian Adrian Grijincu (LP: #761646)
  * Update debian/copyright to copyright specification 1.0
  * Bump standards version to 3.8.3

  [ Harald Sitter ]
  * KDE: don't duplicate entries by improper removal of entries on update
    (LP: #992061)
  * udisks: on successful format the ended callback must be called
  * KDE: strip() ISO labels as some derivates end up having newlines in
    their label (e.g. netrunner)
  * KDE: auto-populate the ISO list with files found in downloads dir
    (LP: #761745)

  [ Jonathan Riddell ]
  * KDE: do not clear formatting list in backend, fix breakage at
    end of format

 -- Harald Sitter <apachelogger@ubuntu.com>  Fri, 06 Jul 2012 13:22:46 +0200

usb-creator (0.2.39) quantal; urgency=low

  [ Colin Watson ]
  * Clean up various pyflakes warnings.
  * Only pass unicode=True to gettext.install in Python 2.
  * Open subprocesses with universal_newlines=True when expecting to read
    text from them.  On Python 2, this only enables \r\n conversion and the
    like, but on Python 3 this also causes subprocess-related file objects
    to read str rather than bytes.
  * Use str() rather than unicode() in Python 3.
  * Remove __pycache__ directories on clean.
  * Use 'isinstance(obj, collections.Callable)' instead of 'callable(obj)'
    in Python 3.
  * Change 'except StandardError' to 'except Exception'; StandardError was
    removed in Python 3.
  * Use Python 3 name for Queue if available.
  * Handle a few cases of builtins being changed to return iterators in
    Python 3.
  * Just use dict.items() rather than bothering with Python 2/3
    compatibility for dict.iteritems().

  [ Evan Dandrea ]
  * Only clear the selected partition on the disk, not the entire disk
    (LP: #484252). Thanks Dmitrijs Ledkovs!

 -- Evan Dandrea <ev@ubuntu.com>  Mon, 28 May 2012 17:58:43 +0100

usb-creator (0.2.38) precise; urgency=low

  [ Colin Watson ]
  * Fix production of Ubuntu <= 10.04 images on Ubuntu >= 10.10 using
    syslinux-legacy (LP: #645818).

  [ Evan Dandrea ]
  * When comparing against 10.04, treat its point releases the same as
    the initial release.
  * Drop code to insert 'ui' in front of 'gfxboot bootlogo' in 10.04.
    The ui command was added after 10.04.

 -- Evan Dandrea <ev@ubuntu.com>  Mon, 16 Apr 2012 11:29:30 +0100

usb-creator (0.2.37) precise; urgency=low

  * debian/control: Fix dependency: python-gobject → python-gi.

 -- Martin Pitt <martin.pitt@ubuntu.com>  Tue, 13 Mar 2012 19:15:49 +0100

usb-creator (0.2.36) precise; urgency=low

  [ Colin Watson ]
  * Use Python 3-compatible print functions.
  * Use "except Exception as e" syntax rather than the old-style "except
    Exception, e".
  * Use "raise Exception(value)" syntax rather than the old-style "raise
    Exception, value".
  * Use absolute imports.

  [ Martin Pitt ]
  * usbcreator/frontends/gtk/frontend.py:  We do not have any help for
    usb-creator, so hide the Help button. (LP: #884008)

 -- Martin Pitt <martin.pitt@ubuntu.com>  Sun, 11 Mar 2012 10:14:12 +0100


usbcreator/backends/base/backend.py
import usbcreator.install
from usbcreator import misc
import logging
import os
import sys

if sys.version_info[0] < 3:
    import ConfigParser as configparser
else:
    import configparser

def abstract(func):
    def not_implemented(*args):
        raise NotImplementedError('%s is not implemented by the backend.' %
                                  func.__name__)
    return not_implemented

class Config:
    def __init__(self, backend):
        self.file = os.path.expanduser('~/.usb-creator.ini')
        self.folder = os.path.expanduser('~')
        self.section = 'images'
        self.parser = configparser.SafeConfigParser()
        self.backend = backend

    def load(self):
        p = self.parser
        p.read(self.file)
        try:
            folder = p.get('DEFAULT', 'folder')
            self.folder = folder
        except:
            python_govno = 'Yes!'
        if p.has_section(self.section):
            for f in p.items(self.section):
                self.backend.add_image(f[1])

    def save(self):
        p = self.parser
        p.set('DEFAULT', 'folder', self.folder)
        s = self.section
        if p.has_section(s):
            p.remove_section(s)
        p.add_section(s)
        n = 0
        for image_file in self.backend.sources.keys():
            p.set(s, str(n), image_file)
        with open(self.file, 'w') as configfile:
            p.write(configfile)

class Backend:
    def __init__(self):
        self.sources = {}
        self.targets = {}
        self.current_source = None
        self.install_thread = None
        self.config = Config(self)

    # Public methods.

    def add_image(self, filename):
        logging.debug('Backend told to add: %s' % filename)
        filename = os.path.abspath(os.path.expanduser(filename))
        if not os.path.isfile(filename):
            return
        if filename in self.sources:
            logging.warn('Source already added.')
            # TODO evand 2009-07-27: Scroll to source and select.
            return

        extension = os.path.splitext(filename)[1]
        # TODO evand 2009-07-25: What's the equivalent of `file` on Windows?
        # Going by extension is a bit rough.
        if not extension:
            logging.error('File did not have an extension.  '
                          'Could not determine file type.')
            # TODO evand 2009-07-26: Error dialog.
            return

        extension = extension.lower()
        if extension == '.iso':
            label = self._is_casper_cd(filename)
            if label:
                self.sources[filename] = {
                    'device' : filename,
                    'size' : os.path.getsize(filename),
                    'label' : label,
                    'type' : misc.SOURCE_ISO,
                }
                if misc.callable(self.source_added_cb):
                    self.source_added_cb(filename)
        elif extension == '.img':
            self.sources[filename] = {
                'device' : filename,
                'size' : os.path.getsize(filename),
                'label' : '',
                'type' : misc.SOURCE_IMG,
            }
            if misc.callable(self.source_added_cb):
                self.source_added_cb(filename)
        else:
            logging.error('Filename extension type not supported.')

    @abstract
    def detect_devices(self):
        pass

    def set_current_source(self, source):
        if source == None or source in self.sources:
            self.current_source = source
        else:
            raise KeyError(source)
        self.update_free()

    def get_current_source(self):
        return self.current_source

    # Common helpers

    def _device_removed(self, device):
        logging.debug('Device has been removed from the system: %s' % device)
        if device in self.sources:
            if misc.callable(self.source_removed_cb):
                self.source_removed_cb(device)
            self.sources.pop(device)
        elif device in self.targets:
            if misc.callable(self.target_removed_cb):
                self.target_removed_cb(device)
            self.targets.pop(device)

    # Signals.

    def source_added_cb(self, drive):
        pass

    def target_added_cb(self, drive):
        pass

    def source_removed_cb(self, drive):
        pass

    def target_removed_cb(self, drive):
        pass

    def target_changed_cb(self, udi):
        pass

    def format_ended_cb(self):
        pass

    def format_failed_cb(self, message):
        pass

    # Installation signals.

    def success_cb(self):
        pass

    def failure_cb(self, message=None):
        pass

    def install_progress_cb(self, complete, remaining, speed):
        pass

    def install_progress_message_cb(self, message):
        pass

    def install_progress_pulse_cb(self):
        pass

    def install_progress_pulse_stop_cb(self):
        pass

    def retry_cb(self, message):
        pass

    def update_free(self):
        if not self.current_source:
            return True

        for k in self.targets:
            status = self.targets[k]['status']
            if status == misc.NEED_FORMAT or status == misc.CANNOT_USE:
                continue
            changed = self._update_free(k)
            if changed and misc.callable(self.target_changed_cb):
                self.target_changed_cb(k)
        return True

    # Internal functions.

    def _update_free(self, k):
        # TODO evand 2009-08-28: Replace this mess with inotify watches.
        # Incorporate accounting for files we will delete.  Defer updates if
        # sufficient time has not passed since the last update.
        if not self.current_source:
            return False
        current_source = self.sources[self.current_source]
        changed = False
        target = self.targets[k]
        free = target['free']
        target['free'] = misc.fs_size(target['mountpoint'])[1]
        if free != target['free']:
            changed = True

        target = self.targets[k]
        status = target['status']
        target['status'] = misc.CAN_USE
        target['persist'] = 0
        if target['capacity'] < current_source['size']:
            target['status'] = misc.CANNOT_USE
        elif target['free'] < current_source['size']:
            target['status'] = misc.NEED_SPACE
        else:
            target['persist'] = (target['free'] - current_source['size'] -
                                 misc.PADDING * 1024 * 1024)
        if status != target['status']:
            changed = True
        # casper cannot handle files larger than MAX_PERSISTENCE (4GB)
        persist_max = misc.MAX_PERSISTENCE * 1024 * 1024 - 1
        if target['persist'] > persist_max:
            target['persist'] = persist_max
        return changed

    def install(self, source, target, persist, device=None,
                allow_system_internal=False):
        fastboot_mode = self.__class__.__name__ == 'FastbootBackend'
        logging.debug('Starting install thread.')
        self.install_thread = usbcreator.install.install(
            source, target, persist, device=device,
            allow_system_internal=allow_system_internal,
            fastboot_mode=fastboot_mode)
        # Connect signals.
        self.install_thread.success = self.success_cb
        self.install_thread.failure = self.failure_cb
        self.install_thread.progress = self.install_progress_cb
        self.install_thread.progress_message = self.install_progress_message_cb
        self.install_thread.progress_pulse = self.install_progress_pulse_cb
        self.install_thread.progress_pulse_stop = self.install_progress_pulse_stop_cb
        self.install_thread.retry = self.retry_cb
        self.install_thread.start()

    def cancel_install(self):
        logging.debug('cancel_install')
        if self.install_thread and self.install_thread.is_alive():
            # TODO evand 2009-07-24: Should set the timeout for join, and
            # continue to process in a loop until the thread finishes, calling
            # into the frontend for UI event processing then break.  The
            # frontend should notify the user that it's canceling by changing
            # the progress message to "Canceling the installation..." and
            # disabiling the cancel button.
            self.install_thread.join()
